reimplement sorting order of controller mappings from 2.0

Be-ing opened the pull request:
... to distinguish them from mappings shipped with Mixxx in the system resource folder. Fixing
https://bugs.launchpad.net/mixxx/+bug/1732712

Mixxx also should not copy the mapping into the user folder unless the user modifies it. However, implementing that would require more extensive changes that I am hesitant to do this far into the 2.1 beta period.

Be-ing committed:
mark mappings from user folder with "(user)"

... to distinguish them from mappings shipped with Mixxx in the system
resource folder. Fixing
https://bugs.launchpad.net/mixxx/+bug/1732712

Be-ing commented:
I'm not totally sure that "(user)" is the best way to mark these. Anyone have other suggestions?

ferranpujolcamins commented:
Many thanks to implement this! Sometimes simple things make people happy :)

Maybe you could do the opposite, mark the mappings shipped with mixxx with
"(preset)". Or something like this.

On 26 Feb 2018 6:36 am, "Be" <notifications@github.com> wrote:

I'm not totally sure that "(user)" is the best way to mark these. Anyone
have other suggestions?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/mixxxdj/mixxx/pull/1524#issuecomment-368395211>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AGIcTxdaW-OpFsX7ZZXL5MP_dpw-ZUdaks5tYkLcgaJpZM4SSoTD>
.

ronso0 commented:
Thanks for this improvement!

> Mixxx also should not copy the mapping into the user folder unless the user modifies it.

I agree, but on the other hand, it allows to back up the whole .mixxx directory and save the current working state with settings, mappings and the database, no matter what changes new versions bring.

Be-ing commented:
>  it allows to back up the whole .mixxx directory and save the current working state with settings, mappings and the database, no matter what changes new versions bring.

I think this benefit is fairly trivial compared to the hassle and confusion the current situation is causing.

Be-ing commented:
>  Maybe you could do the opposite, mark the mappings shipped with mixxx with "(preset)". Or something like this.

Interesting idea, but I think that would be annoying. There are lots of mappings in the system folder but typically only a few in the user folder.

Be-ing commented:
> Mixxx also should not copy the mapping into the user folder unless the user modifies it. However, implementing that would require more extensive changes that I am hesitant to do this far into the 2.1 beta period.

Ugh, I realized this problem will still affect users who install mappings to their user folder, not just developers. I'll try implementing this and change as little as I can...

Be-ing commented:
I won't have time to really fix this issue for 2.1 by implementing bigger and riskier changes to prevent Mixxx from copying the mapping to the local folder unless it was modified. I think we should merge this small improvement as it is.

daschuer commented on src/controllers/controllerpresetinfo.h:
```diff
@@ -48,7 +49,13 @@ class PresetInfo {
+        if (m_name.length() != 0) {
+            return m_name;
+        } else {
+            return QFileInfo(m_path).baseName();
```
When does this happen? Please comment.

daschuer replied:
QFileInfo is quite expensive but includes caching. So a temporary QFileInfo school be avoided.

Be-ing replied:
This is called from enumeratePresets, which is called from DlgPrefController::slotUpdate

daschuer replied:
You should initalize n_name in the constructor. 
You should make sure that only one QFileInfo is used.
This can become a member variable if required, or a local in the Constructor. 

This code place is also redundant now: 
https://github.com/mixxxdj/mixxx/blob/4f1e0832c0a8860efe3d5ab9e4725bebe023e3cc/src/controllers/dlgprefcontroller.cpp#L29

daschuer commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -249,6 +229,15 @@ void DlgPrefController::slotDirty() {
+const QString DlgPrefController::presetDisplayName(const PresetInfo& info) {
+    if (QFileInfo(info.getPath()).absoluteDir() == userPresetsPath(m_pConfig)) {
+        //: Shown in combobox to select controller mappings to distinguish user-installed mappings from mappings included in Mixxx
+        return info.getName() + " (" + tr("user") + ")";
```
User is really not self explaining.
How about (user folder)

daschuer replied:
Or just an asterisk and explain it below.

Be-ing replied:
:/ I don't think either of these are very good options. I am thinking we should revert to the 2.0 behavior, which sorted mappings by file path. This put mappings in the user folder at the top in alphabetical order, followed by the system mappings in alphabetical order. I do not recall any complaints about that in 2.0. That behavior was broken by effbb061a0b0f55c5ece9e5c1df2f5665cca7817 because it removed the QMap of mappings indexed by a QString for the file path.

Be-ing committed:
sort controller mappings in user folder first then system folder

This was how it was done in 2.0. It is not self-explanatory and
kinda clunky, but it's better than the current situation where
they are all sorted alphabetically and users get confused why
they see two identically named mappings next to each other.

This behavior was broken by commit effbb06 because it removed the
QMap of mappings indexed by a QString for the file path.

daschuer commented:
OK, how about to add a fake entry "----" to distingish the local from the user region?

Be-ing commented:
Yes, I thought about that. It will take a little more work though.

daschuer commented:
Do you see a chance to add the line in time or should we merge this now?

Be-ing commented:
I think adding a `---------` entry into the menu would hardly improve the situation. This whole mapping selection UX could use an overhaul. This PR now reimplements the 2.0 behavior. It's not the greatest design, but I don't recall anyone getting too confused about it before it was broken by https://github.com/mixxxdj/mixxx/commit/effbb061a0b0f55c5ece9e5c1df2f5665cca7817 (which seems to have been committed directly to master without review).

daschuer commented:
Ok, so I will merge this now.
Thank you for working in it.
